File: app/src/test/java/com/example/mike/mp3player/client/views/fragments/viewpager/FolderListFragmentTest.java
package com.example.mike.mp3player.client.views.fragments.viewpager;

import android.support.v4.media.MediaBrowserCompat;

import androidx.fragment.app.testing.FragmentScenario;

import com.example.mike.mp3player.client.activities.TestMainActivity;
import com.example.mike.mp3player.client.views.fragments.FragmentTestBase;
import com.example.mike.mp3player.commons.MediaItemType;
import com.example.mike.mp3player.dagger.components.MediaActivityCompatComponent;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.android.controller.ActivityController;

import static com.example.mike.mp3player.TestUtils.createMediaItem;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class FolderListFragmentTest extends FragmentTestBase<FolderListFragment> {
    private static final String FRAGMENT_TAG = "FragmentScenario_Fragment_Tag";

    private FolderListFragment folderListFragment;
    private ActivityController<TestMainActivity> activityScenario;
    @Before
    public void setup() {
        activityScenario = Robolectric.buildActivity(TestMainActivity.class).setup();
        TestMainActivity testMainActivity = activityScenario.get();
        MediaActivityCompatComponent component = testMainActivity.getMediaActivityCompatComponent();
        folderListFragment = new FolderListFragment(MediaItemType.FOLDERS, "id", component);
        super.setup(folderListFragment, FolderListFragment.class);
    }

    @Test
    public void testItemSelected() {
        FragmentScenario.FragmentAction<FolderListFragment> action = this::itemSelected;
        performAction(action);
    }

    private void itemSelected(MediaItemListFragment fragment) {
        MediaItemListFragment spiedFragment = spy(fragment);
        final String id = "ID";
        final String title = "TITLE";
        final String description = "description";

        MediaBrowserCompat.MediaItem mediaItem =
                createMediaItem(id, title, description, MediaItemType.ROOT);
        spiedFragment.itemSelected(mediaItem);
        verify(spiedFragment).startActivity(any());

    }
}

File: android/client/src/testUnitTest/java/com/github/goldy1992/mp3player/client/views/fragments/viewpager/FolderListFragmentTest.kt
package com.github.goldy1992.mp3player.client.views.fragments.viewpager

import android.os.Looper
import androidx.fragment.app.testing.FragmentScenario.FragmentAction
import com.github.goldy1992.mp3player.client.activities.MainActivity
import com.github.goldy1992.mp3player.client.views.fragments.FragmentTestBase
import com.github.goldy1992.mp3player.commons.MediaItemBuilder
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.spy
import com.nhaarman.mockitokotlin2.verify
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import org.robolectric.annotation.LooperMode

@RunWith(RobolectricTestRunner::class)
@LooperMode(LooperMode.Mode.PAUSED)
class FolderListFragmentTest : FragmentTestBase<FolderListFragment>() {

    private var folderListFragment: FolderListFragment? = null

    @Before
    fun setup() {
        activityScenario = Robolectric.buildActivity(MainActivity::class.java)
                .setup()
        val testMainActivity: MainActivity = activityScenario!!.get()
        val component = testMainActivity.mediaActivityCompatComponent
        folderListFragment = FolderListFragment.newInstance(MediaItemType.FOLDERS, "id")
        super.setup(folderListFragment, FolderListFragment::class.java)
    }

    @Test
    fun testItemSelected() {
        val action = FragmentAction<FolderListFragment> { fragment: FolderListFragment? -> itemSelected(fragment) }
        performAction(action)
    }

    private fun itemSelected(fragment: MediaItemListFragment?) {
        val spiedFragment = spy(fragment)

        val id = "ID"
        val title = "TITLE"
        val description = "description"
        val mediaItem = MediaItemBuilder(id)
                 .setTitle(title)
                 .setDescription(description)
                 .build()
        spiedFragment!!.itemSelected(mediaItem)
        Shadows.shadowOf(Looper.getMainLooper()).idle()
        verify(spiedFragment).startActivity(any())
    }

    companion object {
        private const val FRAGMENT_TAG = "FragmentScenario_Fragment_Tag"
    }
}